### PR TITLE
Add fading sounds

### PR DESCRIPTION
**Pull request by Bremaweb**
Adds fading sounds in and out to the modding api

### CODE REVIEW

**Comment by MarkuBu** [comment]
#4304

File: src/server.cpp [hunk 1]
```diff
@@ -2052,6 +2052,27 @@ void Server::stopSound(s32 handle)
...
+	std::map<s32, ServerPlayingSound>::iterator i =
+			m_playing_sounds.find(handle);
+	if(i == m_playing_sounds.end())
+		return;
+	ServerPlayingSound &psound = i->second;
+
+	NetworkPacket pkt(TOCLIENT_FADE_SOUND, 4);
+	pkt << handle << step << gain;
+
+	for(std::set<u16>::iterator i = psound.clients.begin();
```
Comment (nerzhul) [inline src/server.cpp]: missing space

File: src/network/clientpackethandler.cpp [hunk 2]
```diff
@@ -773,14 +773,15 @@ void Client::handleCommand_PlaySound(NetworkPacket* pkt)
 	v3f pos;
 	u16 object_id;
 	bool loop;
+	float fade;
 
-	*pkt >> server_id >> name >> gain >> type >> pos >> object_id >> loop;
+	*pkt >> server_id >> name >> gain >> type >> pos >> object_id >> loop >> fade;
```
Comment (nerzhul) [inline src/network/clientpackethandler.cpp]: you should be retrocompatible, use a try/catch directive to handle the missing fade

File: src/sound_openal.cpp [hunk 3]
```diff
@@ -273,6 +273,19 @@ class OpenALSoundManager: public ISoundManager
 	std::map<std::string, std::vector<SoundBuffer*> > m_buffers;
 	std::map<int, PlayingSound*> m_sounds_playing;
 	v3f m_listener_pos;
+	struct fade_status {
```
Comment (SmallJoker) [inline src/sound_openal.cpp]: Structs should be `PascalCase`

File: src/script/lua_api/l_util.h [hunk 4]
```diff
@@ -21,7 +21,10 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #define L_UTIL_H_
 
 #include "lua_api/l_base.h"
+#include "config.h"
 
+struct HTTPFetchRequest;
+struct HTTPFetchResult;
```
Comment (SmallJoker) [inline src/script/lua_api/l_util.h]: I don't see any place where they are used in `l_util`. Is this change required?

Comment (Bremaweb) [inline src/script/lua_api/l_util.h]: I think they were left over from trying out pull requests. I'll remove them.

File: src/sound_openal.cpp [hunk 5]
```diff
@@ -593,6 +656,26 @@ class OpenALSoundManager: public ISoundManager
...
+		std::map<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
+		if(i == m_sounds_playing.end())
+			return false;
+
+		PlayingSound *sound = i->second;
+		alSourcef(sound->source_id, AL_GAIN, gain);
+		return true;
+	}
+
+	float getSoundGain(int id){
```
Comment (SmallJoker) [inline src/sound_openal.cpp]: For all functions: New line after `)`

File: src/sound_openal.cpp [hunk 6]
```diff
@@ -593,6 +656,26 @@ class OpenALSoundManager: public ISoundManager
...
+		std::map<int, PlayingSound*>::iterator i = m_sounds_playing.find(id);
+			if(i == m_sounds_playing.end())
+				return 0;
```
Comment (SmallJoker) [inline src/sound_openal.cpp]: One indent too much on both lines.

File: src/network/clientpackethandler.cpp [hunk 7]
```diff
@@ -773,14 +773,19 @@ void Client::handleCommand_PlaySound(NetworkPacket* pkt)
 	v3f pos;
 	u16 object_id;
 	bool loop;
+	float fade = 0;
 
-	*pkt >> server_id >> name >> gain >> type >> pos >> object_id >> loop;
+	try	{
+		*pkt >> server_id >> name >> gain >> type >> pos >> object_id >> loop >> fade;
+	} catch (...) {
+		*pkt >> server_id >> name >> gain >> type >> pos >> object_id >> loop;
```
Comment (nerzhul) [inline src/network/clientpackethandler.cpp]: this catch is absolutely wrong because you already read all values. Please let an empty catch.

File: src/script/lua_api/l_util.h [hunk 8]
```diff
@@ -21,6 +21,7 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 #define L_UTIL_H_
 
 #include "lua_api/l_base.h"
+#include "config.h"
```
Comment (nerzhul) [inline src/script/lua_api/l_util.h]: why is this included ?

File: src/network/clientopcodes.cpp [hunk 9]
```diff
@@ -108,7 +108,7 @@ const ToClientCommandHandler toClientCommandTable[TOCLIENT_NUM_MSG_TYPES] =
 	{ "TOCLIENT_LOCAL_PLAYER_ANIMATIONS",  TOCLIENT_STATE_CONNECTED, &Client::handleCommand_LocalPlayerAnimations }, // 0x51
 	{ "TOCLIENT_EYE_OFFSET",               TOCLIENT_STATE_CONNECTED, &Client::handleCommand_EyeOffset }, // 0x52
 	{ "TOCLIENT_DELETE_PARTICLESPAWNER",   TOCLIENT_STATE_CONNECTED, &Client::handleCommand_DeleteParticleSpawner }, // 0x53
-	null_command_handler,
+	{ "TOCLIENT_FADE_SOUND",			   TOCLIENT_STATE_CONNECTED, &Client::handleCommand_FadeSound }, // 0x54
```
Comment (SmallJoker) [inline src/network/clientopcodes.cpp]: Use spaces to indent after the first letter on a line.

File: src/network/clientpackethandler.cpp [hunk 10]
```diff
@@ -821,6 +826,26 @@ void Client::handleCommand_StopSound(NetworkPacket* pkt)
 	}
 }
 
+void Client::handleCommand_FadeSound(NetworkPacket* pkt)
```
Comment (SmallJoker) [inline src/network/clientpackethandler.cpp]: Pointer on the wrong side, they belong to the variable name. `NetworkPacket *pkt`

File: src/sound_openal.cpp [hunk 11]
```diff
@@ -575,6 +599,47 @@ class OpenALSoundManager: public ISoundManager
...
+		float cGain = getSoundGain(soundid);
+		FadeState f = FadeState(step,cGain,gain);
+		m_sounds_fading[soundid] = f;
+	}
+
+	void doFades(float dtime)
+	{
+		static float fadeDelay = 0;
+		fadeDelay += dtime;
+		if ( fadeDelay > 0.2 ){
```
Comment (SmallJoker) [inline src/sound_openal.cpp]: Please make sure that all your if-checks are in the following format: `if (number > number) {`. No matter if the code around is formatted wrongly.

File: src/sound_openal.cpp [hunk 12]
```diff
@@ -575,6 +599,47 @@ class OpenALSoundManager: public ISoundManager
...
+			float chkGain = 0;
+			for ( m_fading_it i = m_sounds_fading.begin(); i != m_sounds_fading.end(); i++ ){
+				if ( i->second.step < 0 )
+					chkGain = i->second.current_gain * -1;
+				else
+					chkGain = i->second.current_gain;
+
+				if ( chkGain < i->second.target_gain ){
+					i->second.current_gain += (i->second.step * fadeDelay);
+					if ( i->second.current_gain < 0 )
```
Comment (SmallJoker) [inline src/sound_openal.cpp]: `i->second.current_gain = rangelim(i->second.current_gain, 0, 1);`

**Comment by sofar** [comment]
Would *really* appreciate if this PR could get updated!

**Comment by SmallJoker** [comment]
Adapted in #5340.
